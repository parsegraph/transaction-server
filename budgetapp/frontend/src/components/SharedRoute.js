import React from 'react';
import {connect, useDispatch } from 'react-redux';
import { Route, Redirect } from 'react-router-dom';
import {hasUserSession} from '../redux/actions/signUpActions';

function SharedRoute({ component: Component, userComponent: UserComponent, roles, ...rest }) {
    const dispatch = useDispatch();
    dispatch(hasUserSession());

    const hasSession = rest.signup.hasUserSession;
    return (
        <Route {...rest} render={props => {
            if (!hasSession) {
                // not logged in so redirect to anonymous component
                return <Component {...props}/>;
            }

            // logged in so return user's component
            return <UserComponent {...props} />
        }} />
    );
}

const mapStateToProps = (state) => {
    return {signup:state.signup}
}

SharedRoute = connect(mapStateToProps)(SharedRoute);
export {SharedRoute};