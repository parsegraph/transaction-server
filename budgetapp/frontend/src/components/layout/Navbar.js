
import "../layout/styles/sharedstyles.css";

