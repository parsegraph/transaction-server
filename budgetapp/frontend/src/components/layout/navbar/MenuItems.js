import React from "react";
import { logout } from "../../../redux/actions/signUpActions";
import { Link } from 'react-router-dom';
import { Button } from "./Button";

export const MenuItems = [
  {
    title: "Home",
    url: "/",
    cName: "navbar-links",
  },
  {
    title: "Accounts",
    url: "/account",
    cName: "navbar-links",
    userOnly: true,
  },
  {
    title: "Log in",
    url: "/login",
    cName: "navbar-links",
    anonymousOnly: true,
  },
  {
    title: "Logout",
    url: "/api/auth/logout",
    cName: "navbar-links",
    userOnly: true,
    func: (item, index, logout) => {return <li key={index}>
    <a className={item.cName} onClick={logout}>
      {item.title}
    </a>
  </li>}
  },
  {
    title: "Sign up",
    anonymousOnly: true,
    func: (item, index) => {
      return (
        <li key={index}>
          <Link to="/signup"><Button className="nav-signup">Sign Up</Button></Link>
        </li>
      );
    }
  }
];
