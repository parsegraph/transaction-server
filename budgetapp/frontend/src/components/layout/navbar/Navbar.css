.NavbarItems {
  background: rgb(50, 175, 152);
  height: 80px;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: 1.2rem;
}

.navbar-logo {
  color: #fff;
  justify-self: start;
  cursor: pointer;
}

.fa-dollar-sign {
  margin-left: 0.5rem;
  font-size: 1.6rem;
}

.nav-menu {
  display: grid;
  grid-template-columns: repeat(5, auto);
  grid-gap: 10px;
  list-style: none;
  text-align: center;
  width: 70vw;
  justify-content: end;
  margin-right: 2rem;
  align-items: center;
  margin-bottom: 0;
}

.navbar-links {
  color: white;
  text-decoration: none;
  padding: 0.5rem 1rem;
}

#nav-signup {
  margin: 3rem;
}

.navbar-links:hover {
  background-color: #6d76f6;
  border-radius: 4px;
  transition: all 0.2s ease-out;
}

.fa-bars {
  color: #fff;
}

.menu-icon {
  display: none;
}

.navbar-logo {
  margin-bottom: 0;
}

@media screen and (max-width: 960px) {
  .NavbarItems {
    position: relative;
  }

  .nav-menu {
    display: flex;
    flex-direction: column;
    width: 100%;
    position: absolute;
    top: 80px;
    left: -100%;
    opacity: 1;
    transition: all 0.5s ease;
  }

  .nav-menu.active {
    background: #6668f4;
    left: 0;
    opacity: 1;
    transition: all 0.5s ease;
    z-index: 1;
  }

  .navbar-links {
    text-align: center;
    padding: 3rem;
    width: 100%;
    display: table;
  }

  .navbar-links:hover {
    background-color: #7577fa;
    border-radius: 0;
  }

  .navbar-logo {
    position: absolute;
    left: 1em;
  }

  .menu-icon {
    display: block;
    position: absolute;
    right: 1em;
    font-size: 1.8rem;
    cursor: pointer;
  }

  .fa-times {
    color: #fff;
    font-size: 2rem;
  }

  .navbar-links-mobile {
    display: block;
    text-align: center;
    padding: 1.5rem;
    margin: 2rem auto;
    border-radius: 4px;
    width: 80%;
    background: #4ad9e4;
    text-decoration: none;
    color: white;
    font-size: 1.5rem;
  }

  .navbar-links-mobile:hover {
    background: white;
    color: #6568f4;
    transition: 250ms;
  }
}
