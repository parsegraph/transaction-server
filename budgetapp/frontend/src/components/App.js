import React, {useReducer} from "react";
import ReactDOM from "react-dom";
import { BrowserRouter as Router, Route, Switch } from "react-router-dom";

import { Navbar } from "./layout/navbar/Navbar";
import Login from ".//layout/pages/Login";
import Accounts from "./layout/pages/Accounts";
import Transactions from "./layout/pages/Transactions";
import "./layout/styles/App.css";
import SignUp from "./layout/pages/SignUp";
import api from '../requests';

function App() {
  const [login] = useReducer(api.loginReducer, {});

  if (login.user) {
    return (
      <Router>
        <Navbar />
        <main>
          <Switch>
            <Route exact strict path="/" component={Accounts} />
            <Route exact path="/transaction" component={Transactions} />
          </Switch>
        </main>
      </Router>
    );
  }

  return (
    <Router>
      <Navbar />
      <main>
        <Switch>
          <Route exact strict path="/" component={SignUp} />
          <Route exact strict path="/auth/user" component={Login} />
        </Switch>
      </main>
    </Router>
  );
}

ReactDOM.render(<App />, document.getElementById("app"));
