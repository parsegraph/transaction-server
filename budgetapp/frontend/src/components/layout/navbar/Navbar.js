import React, { Component } from "react";
import { MenuItems } from "./MenuItems";
import { Button } from "./Button";
import "./Navbar.css";
import { connect } from "react-redux";
import { logout } from "../../../redux/actions/signUpActions";

class Navbar extends Component {
  
  state = { clicked: false };
  
  handleClick = () => {
    this.setState({ clicked: !this.state.clicked });
  };
  render() {
    return (
      <nav className="NavbarItems">
        <h1 className="navbar-logo">
          <i className="fas fa-cash-register"></i>
        </h1>
        <div className="menu-icon" onClick={this.handleClick}>
          <i
            className={this.state.clicked ? "fas fa-times" : "fas fa-bars"}
          ></i>
        </div>
        <ul className={this.state.clicked ? "nav-menu active" : "nav-menu"}>
          {MenuItems.map((item, index) => {
            if(item.func) {
              return item.func(item, index, this.props.logout);
            }
            return (
                <li key={index}>
                  <a className={item.cName} href={item.url}>
                    {item.title}
                  </a>
                </li>
            );
          })}
          <li>
            <Button className="nav-signup">Sign Up</Button>
          </li>
        </ul>
      </nav>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
    return {
        logout: () => dispatch(logout())
    }
};
export default connect(null, mapDispatchToProps)(Navbar);
