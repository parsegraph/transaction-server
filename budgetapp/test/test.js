import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider, useDispatch as useDispatchMock } from 'react-redux';
import React, { useState as useStateMock } from 'react';
import configureMockStore from 'redux-mock-store';

import Navbar from '../frontend/src/components/layout/navbar/Navbar';
import {Button} from '../frontend/src/components/layout/navbar/Button';

import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';
import renderer from 'react-test-renderer';
import { SharedRoute } from '../frontend/src/components/SharedRoute';
import { Route, MemoryRouter, Switch } from 'react-router-dom';

configure({adapter: new Adapter()});

const mockStore = configureMockStore([thunk]);

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}));

describe("Navbar", ()=>{
  let wrapper;
  const createWrapper = (conf)=>{
    const store = mockStore(conf || {});
    return mount(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar/>
        </MemoryRouter>
      </Provider>
    );
  };

  it('renders correctly', () => {
    wrapper = createWrapper({signup:{hasUserSession:false}});
    const tree = renderer
      .create(wrapper)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show Sign Up when NOT logged in', ()=>{
    wrapper = createWrapper({signup:{hasUserSession:false}});
    expect(wrapper.containsMatchingElement(<Button className="nav-signup">Sign Up</Button>)).toEqual(true);
  });

  it('should NOT show Sign Up when logged in', ()=>{
    wrapper = createWrapper({signup:{hasUserSession:true}});
    expect(wrapper.containsMatchingElement(<Button className="nav-signup">Sign Up</Button>)).toEqual(false);
  });
});

describe("SharedRoute", ()=>{
  const createWrapper = (conf)=>{
    const store = mockStore(conf);
    return mount(
      <Provider store={store}>
        <MemoryRouter>
          <Switch>
            <SharedRoute path='*' userComponent={()=>(<Button>user</Button>)} component={()=>(<Button>anonymous</Button>)}/>
          </Switch>
        </MemoryRouter>
      </Provider>
    );
  };

  it('renders correctly when anonymous', () => {
    let wrapper = createWrapper({signup:{hasUserSession:false}});
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when logged in', () => {
    let wrapper = createWrapper({signup:{hasUserSession:true}});
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show up as user if logged in', ()=>{
    let wrapper = createWrapper({signup:{hasUserSession:true}});
    expect(wrapper.containsMatchingElement(<Button>user</Button>)).toEqual(true)
  });

  it('should show up as anonymous if not logged in', ()=>{
    let wrapper = createWrapper({signup:{hasUserSession:false}});
    expect(wrapper.containsMatchingElement(<Button>anonymous</Button>)).toEqual(true)
  });
});