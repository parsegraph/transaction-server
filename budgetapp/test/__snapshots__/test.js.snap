// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<nav
  className="NavbarItems"
>
  <h1
    className="navbar-logo"
  >
    <i
      className="fas fa-cash-register"
    />
  </h1>
  <div
    className="menu-icon"
    onClick={[Function]}
  >
    <i
      className="fas fa-bars"
    />
  </div>
  <ul
    className="nav-menu"
  >
    <li>
      <a
        className="navbar-links"
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
    </li>
    <li>
      <a
        className="navbar-links"
        href="/login"
        onClick={[Function]}
      >
        Log in
      </a>
    </li>
    <li>
      <a
        href="/signup"
        onClick={[Function]}
      >
        <button
          className="btn btn--primary btn--medium"
          id="nav-signup"
        >
          Sign Up
        </button>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`SharedRoute renders correctly when anonymous 1`] = `
<button
  className="btn btn--primary btn--medium"
  id="undefined"
>
  anonymous
</button>
`;

exports[`SharedRoute renders correctly when logged in 1`] = `
<button
  className="btn btn--primary btn--medium"
  id="undefined"
>
  user
</button>
`;
